Userland/Libraries/LibCodeComprehension/Types.h

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2022, Itamar S. <itamar8910@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/String.h>
+
+#pragma once
+
+namespace CodeComprehension {
+
+enum class Language {
+    Unspecified,
+    Cpp,
+};
+
+struct AutocompleteResultEntry {
+    String completion;
+    size_t partial_input_length { 0 };
+    Language language { Language::Unspecified };
+    String display_text {};
+
+    enum class HideAutocompleteAfterApplying {
+        No,
+        Yes,
+    };
+    HideAutocompleteAfterApplying hide_autocomplete_after_applying { HideAutocompleteAfterApplying::Yes };
+};
+
+struct ProjectLocation {
+    String file;
+    size_t line { 0 };
+    size_t column { 0 };
+
+    bool operator==(ProjectLocation const& other) const
+    {
+        return file == other.file && line == other.line && column == other.column;
+    }
+};
+
+enum class DeclarationType {
+    Function,
+    Struct,
+    Class,
+    Variable,
+    PreprocessorDefinition,
+    Namespace,
+    Member,
+};
+
+struct Declaration {
+    String name;
+    ProjectLocation position;
+    DeclarationType type;
+    String scope;
+
+    bool operator==(Declaration const& other) const
+    {
+        return name == other.name && position == other.position && type == other.type && scope == other.scope;
+    }
+};
+
+#define FOR_EACH_SEMANTIC_TYPE        \
```

Comment:
Let's have the other macro as an argument to this one; I think that pattern is more common.

Reply:
Not sure I understand what you meant here. Can you point at an example in the repo for doing it the other way around?

Reply:
okay nvm